stl/inc/bit

```diff
@@ -148,19 +139,59 @@ _NODISCARD int _Checked_x86_x64_countl_zero(const _Ty _Val) noexcept {
         return static_cast<int>(__lzcnt(_Val));
     } else {
 #ifdef _M_IX86
+        static_assert(_Digits <= 32, "Should have handled this in _Checked_x86_x64_countl_zero");
+#else // ^^^ _M_IX86 / !_M_IX86 vvv
+        return static_cast<int>(__lzcnt64(_Val));
+#endif // _M_IX86
+    }
+    // note: we don't need to call a fallback here because
+    // all supported x86 processors at least have bsr/bsf
+}
+
+template <class _Ty>
+_NODISCARD int _Checked_x86_x64_countl_zero(const _Ty _Val) noexcept {
+    if constexpr (sizeof(_Ty) > sizeof(void*)) {
         const unsigned int _High = _Val >> 32;
         const auto _Low          = static_cast<unsigned int>(_Val);
         if (_High == 0) {
             return 32 + _Checked_x86_x64_countl_zero(_Low);
         } else {
             return _Checked_x86_x64_countl_zero(_High);
         }
-#else // ^^^ _M_IX86 / !_M_IX86 vvv
-        return static_cast<int>(__lzcnt64(_Val));
-#endif // _M_IX86
+    } else {
+        int _Result                            = _Unchecked_x86_x64_countl_zero(_Val);
+#ifndef __AVX2__
+        static constexpr char _Have_lzcnt      = 0;
+        static constexpr char _Dont_Have_lzcnt = 1;
+        static constexpr char _Deteting_lzcnt  = 2;
+
+        static char _Lzcnt_presence = _Deteting_lzcnt;
+        // lzcnt (when it doesn't fall back to bsr) is defined correctly for zero
+        // bsr has undefined output for zero
+        auto _Lzcnt_presence_local = __iso_volatile_load8(&_Lzcnt_presence);
+        for (;;) {
+            if (_Lzcnt_presence_local == _Have_lzcnt) {
+                return _Result;
+            } else if (_Lzcnt_presence_local == _Dont_Have_lzcnt) {
+                if (_Val == 0) {
+                    return 0;
+                } else {
+                    constexpr int _Digits        = numeric_limits<_Ty>::digits;
+                    constexpr int _Digits_of_bsr = (_Digits < 16) ? 16 : _Digits;
+                    // bsr counts from least significant bit, lzcnt counts from most significant bit, apply correction
+                    return _Digits_of_bsr - 1 - _Result;
+                }
+            }
+
+            volatile unsigned _Test        = 0x8000'0000;
+            volatile unsigned _Test_result = __lzcnt(_Test);
+            _Lzcnt_presence_local          = ((_Test_result == 0) ? _Have_lzcnt : _Dont_Have_lzcnt);
+            __iso_volatile_store8(&_Lzcnt_presence, _Lzcnt_presence_local);

```

Comment:
This mechanism is clever (simply observe the behavior of the instruction, and cache the result), but it involves a lot of code, and I wonder whether it's necessary. The main complexity is that lzcnt support doesn't exactly correspond to `__isa_available`, which is determined in vcstartup. However, what about the following pattern:
* If `__AVX2__` is defined (i.e. we're compiling with `/arch:AVX2` or higher), unconditionally use lzcnt, since AVX2 support should imply lzcnt support (although the converse doesn't always hold, if I understand it correctly).
* Otherwise, check `const bool _Have_lzcnt = __isa_available >= __ISA_AVAILABLE_AVX2;` at runtime. If that's true, use `lzcnt`. If called in a loop, this branch should be correctly predicted, so it should be near-free.
* If that's false, then use the old `_BitScanReverse` intrinsic, special-case zero, and adjust the value appropriately.

This will lead to using `_BitScanReverse` on systems where `lzcnt` is physically supported, but not all of AVX2. I think that's acceptable - processors began supporting AVX2 in 2013-2015, so pre-AVX2 machines are getting old enough to be replaced now. It seems unlikely that the relatively small performance difference between BSR and LZCNT will be problematic, given that (1) a programmer-user will need to upgrade to VS 2019 16.8+, (2) compile with `/std:c++latest` (still "experimental"!) and use `<bit>`, and (3) ship code to end users that are still running this specific set of old CPUs. (The `__lzcnt` processor intrinsic remains available for such programmer-users to directly access.)

Reply:
So should it be
```
#ifdef __AVX2__
__lzcnt(...);
#else
if (__isa_available >= __ISA_AVAILABLE_AVX2) {
   __lzcnt(...);
} else {
   __BitScanReverse(...);
}
```
or just
```
#ifdef __AVX2__
__lzcnt(...);
#else
__BitScanReverse(...);
```
?

Reply:
Read it again, yes I see will have runtime branch too